### PR TITLE
fix EventArgs::getEntityManger deprecation

### PR DESCRIPTION
**Pull request by DubbleClick**
tackles #2638 and #2640

### CODE REVIEW

File: doc/symfony6.md
Comment (DubbleClick) [inline doc/symfony6.md:170]: This isn't optimal, the correct way to go about this would be to stop inheriting from MappedEventSubscriber in https://github.com/doctrine-extensions/DoctrineExtensions/blob/main/src/AbstractTrackingListener.php and changing the code entirely, but that's unfortunately a little past my familiarity with the project.

File: src/Uploadable/Event/UploadableBaseEventArgs.php
Comment (DubbleClick) [inline src/Uploadable/Event/UploadableBaseEventArgs.php:105]: I decided against firing a deprecation here as an EntityManagerInterface is indeed injected

**Comment by DubbleClick** [comment]
@franmomu please trigger the workflows again, getEntity() should no longer be called by us.

**Comment by DubbleClick** [comment]
It is however noteworthy that there is use of plenty other deprecated functionality throughout the project (EntityManager::create) for example. Not all of it is trivial to work around and that unfortunately exceeds the time I can currently invest.

File: doc/symfony6.md
Comment (phansys) [inline doc/symfony6.md:1]: How is the integration with Symfony 6 related to the deprecation fix for `EventArgs::getEntityManger()`?

Comment (DubbleClick) [reply doc/symfony6.md:1]: It's not related to EntityManager, it's related to the deprecation of Event Subscribers. This comes from the Symfony Bridge: https://github.com/symfony/symfony/issues/49586

**Comment by franmomu** [comment]
I think we are doing two things here, apart from avoiding calls to `getEntityManager()` from doctrine, we are also deprecating some of our methods, I would leave deprecating these methods for another PR (which would require a minor release)

**Comment by DubbleClick** [comment]
I'll be happy to remove the deprecation of the methods and open another PR for that. Should I?

**Comment by franmomu** [comment]
I was thinking about adding some fixes like this one and https://github.com/doctrine-extensions/DoctrineExtensions/pull/2651 at least and make a patch release and then make a minor one adding deprecations + removing support for php < 7.4 and probably symfony < 5.4 cc @phansys

**Comment by DubbleClick** [comment]
and #2634 for the holy trinity of pull requests to get rid of all deprecations for now

**Comment by franmomu** [comment]
@DubbleClick if you could please split this PR into two, one that fixes triggering deprecations (could be this one) and another one deprecating methods. 

The `symfony6.md` file won't be necessary, we'll deal with that in https://github.com/doctrine-extensions/DoctrineExtensions/pull/2667 (thanks for creating the file).

**Comment by DubbleClick** [comment]
yep will do it tomorrow

**Comment by DubbleClick** [comment]
@franmomu @phansys sorry, took a little longer

I removed all the deprecations and only kept in the added methods (and a safety check for $platform)

**Comment by DubbleClick** [comment]
rebased it all to only show up in one commit and also created one for EntityManager::create deprecations at https://github.com/doctrine-extensions/DoctrineExtensions/pull/2669

**Comment by DubbleClick** [comment]
I'm not entirely sure how to get rid of all the warnings. Either there will be a warning that ObjectManager is returned rather than EntityManagerInterface, or there will be warnings that ObjectManager doesn't have methods like createQueryBuilder.

**Review by franmomu — COMMENTED** [review]
Thanks for keep working on this, sorry about reverting changes, but it's better if in the same commit there are only changes related to the current fix.

Sorry that I wasn't clear 😞, what I meant in https://github.com/doctrine-extensions/DoctrineExtensions/pull/2639#issuecomment-1677277868 was to only add the necessary code to not trigger deprecations.

I think adding `ORM::setObjectManager`, `UploadableBaseEventArgs::getObjectManager()` and `UploadableBaseEventArgs::getObject()` makes sense if we deprecate the old ones, so I would leave that to the other PR.

File: src/Translatable/Query/TreeWalker/TranslationWalker.php
Comment (DubbleClick) [reply src/Translatable/Query/TreeWalker/TranslationWalker.php]: Time to run
https://cs.symfony.com/doc/rules/return_notation/return_assignment.html
and
https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#singleinarraytocomparerector

then, but I'll leave that to you :)

File: src/SoftDeleteable/Filter/SoftDeleteableFilter.php
Comment (DubbleClick) [reply src/SoftDeleteable/Filter/SoftDeleteableFilter.php]: I've reverted it, but it certainly can be:
image

**Comment by DubbleClick** [comment]
> Sorry that I wasn't clear 😞, what I meant in [#2639 (comment)](https://github.com/doctrine-extensions/DoctrineExtensions/pull/2639#issuecomment-1677277868) was to only add the necessary code to not trigger deprecations.

No worries, I've removed everything not strictly necessary in order to not trigger deprecations now.

> I think adding `ORM::setObjectManager`, `UploadableBaseEventArgs::getObjectManager()` and `UploadableBaseEventArgs::getObject()` makes sense if we deprecate the old ones, so I would leave that to the other PR.

It'll be inevitable at some point, but it doesn't have to happen right now.

File: tests/Gedmo/Mapping/MappingEventAdapterTest.php
Comment (franmomu) [inline tests/Gedmo/Mapping/MappingEventAdapterTest.php:15]: There is this code 😞 

https://github.com/doctrine-extensions/DoctrineExtensions/blob/351caed07e47552ab0df63cafb4fdcccc009eb59/src/Mapping/MappedEventSubscriber.php#L252-L277

that do some funny things to determine which event adapter has to be instantiated and it's based on the event class namespace that should include `ORM` or `ODM`.

It would be awesome to be able to get rid of this code, but since it doesn't seem easy, maybe we should keep using the deprecated `Doctrine\Persistence\Event\LifecycleEventArgs;`, once https://github.com/doctrine-extensions/DoctrineExtensions/pull/2669 is merged we can use one of the new dedicated events.

File: src/SoftDeleteable/Filter/SoftDeleteableFilter.php
Comment (franmomu) [reply src/SoftDeleteable/Filter/SoftDeleteableFilter.php]: That's weird, the implementation is:

https://github.com/doctrine/dbal/blob/63646ffd71d1676d2f747f871be31b7e921c7864/src/Connection.php#L306-L321

and also in the information says it returns a `AbstractPlatform`, don't know why it says `AbstractPlatform|null` on top 🤷‍♂️

File: tests/Gedmo/Mapping/MappingEventAdapterTest.php
Comment (DubbleClick) [reply tests/Gedmo/Mapping/MappingEventAdapterTest.php]: GitHub doesn't have the correct emoji to express my feelings about that code. Sometimes I feel like code reflection goes just a little too far in some languages, hahaha.

Anyway, I've reverted back to Doctrine\ORM\Event\LifecycleEventArgs.

**Review by franmomu — APPROVED** [review]
I see this fine after adding the changelog, I'll approve since I'm not sure if this week I'll be available

File: src/Mapping/Event/Adapter/ORM.php
Comment (phansys) [reply src/Mapping/Event/Adapter/ORM.php:98]: I'm asking about the implementation provided by this PR, in order to determine if a `@todo` comment with the instructions for the next major release is required.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: once #2669 is merged this check becomes obsolete

File: doc/symfony6.md
Comment (stof) [reply doc/symfony6.md]: The simpler solution here might be to make the documentation link to the bundle at https://packagist.org/packages/stof/doctrine-extensions-bundle, which handles registering listeners instead of subscribers, with CI checks warning us if the events used by a listener are updated.

File: src/Mapping/Event/Adapter/ORM.php
Comment (stof) [reply src/Mapping/Event/Adapter/ORM.php:98]: Given that all versions before 2.14.x are EOL (even 2.14.x is), I would suggest simplifying the implementation by bumping the lower bound of the constraint.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: +1, will change this

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php]: @phansys did a version bump to 2.14 and removed the \method_exists check in another commit. If you wish to keep the version at 2.10 for now, just merge the first commit.

Comment (franmomu) [reply src/Mapping/Event/Adapter/ORM.php]: Sorry for taking so long, I think the `\method_exists` will still be necessary, we are working with `EventArgs`: https://github.com/doctrine/event-manager/blob/12f08bd277ce61d7683253264c1d58b797841604/src/EventArgs.php which doesn't have `getObject()` nor `getObjectManager()`.

Custom implementations like `UploadableBaseEventArgs` doesn't have these methods (we should add these methods and deprecate `getEntity` and `getEntityManager` as discussed before).

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php]: Do you want me to undo the last commit, or add these necessary changes back in?

Comment (franmomu) [reply src/Mapping/Event/Adapter/ORM.php]: IMO the last commit should be reverted, I've created https://github.com/doctrine-extensions/DoctrineExtensions/pull/2697 for the conflict `<2.14`.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php]: Will do tomorrow.

Comment (phansys) [reply src/Mapping/Event/Adapter/ORM.php:98]: I think we need a `@todo` comment explaining what to do with these checks when dropping the support for the versions that don't are declaring these methods.

Comment (stof) [reply src/Mapping/Event/Adapter/ORM.php:98]: to me, the way to handle custom implementations provided inside the doctrine-extensions package is to add the methods in those classes (the old methods can then be deprecated).

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: That's what I did at first, haha.

File: src/Mapping/Event/Adapter/ORM.php
Comment (phansys) [inline src/Mapping/Event/Adapter/ORM.php:98]: Could you please use the same approach than #2547, creating and commenting the method, and adding the `@method` annotation at `UploadableBaseEventArgs`?
https://github.com/doctrine-extensions/DoctrineExtensions/blob/c8a995317bb0a3de88f64f912c8d2ca7b76095e4/src/Tree/Node.php#L12-L54

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: @phansys sure I could do that, but at that point, what's the point in even keeping the check in place?
Current master already restricts doctrine/orm to 2.14+, when the method is implemented in UploadableBaseEventArgs, it will always exist.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: I think the better way to go about it is to create the method, change the pipeline to ignore getObjectManager calls on EventArgs (it currently ignores getEventManager calls) and get rid of the check. It doesn't actually hurt backward compatibility in any way, or am I missing something?

Comment (phansys) [reply src/Mapping/Event/Adapter/ORM.php:98]: It sounds logic at first sight.
Let me check out the PR locally to have a better understanding of the scenario here, as I'm not following this topic.

Comment (phansys) [reply src/Mapping/Event/Adapter/ORM.php:98]: From what I see, the case seems to be even worse than I thought, as the property used here (`ORM::$args`) is using a type (`Doctrine\Common\EventArgs`) that does not guarantee the `getEntityManager()` or the `getEntityManager()` methods:
https://github.com/doctrine-extensions/DoctrineExtensions/blob/c8a995317bb0a3de88f64f912c8d2ca7b76095e4/src/Mapping/Event/Adapter/ORM.php#L52

AFAIK, the only usable class declaring these methods is [`Doctrine\Persistence\Event\LifecycleEventArgs`](
https://github.com/doctrine/persistence/blob/8a710c21dfe6e88df21c371012be1bf6f2337525/src/Persistence/Event/LifecycleEventArgs.php#L16), as the [implementation for the ORM](https://github.com/doctrine/orm/blob/9f555ea8fb4f7ae9a362285dc71e85776a1c5cc3/lib/Doctrine/ORM/Event/LifecycleEventArgs.php#L19) is deprecated.

That said, I know we surely couldn't use `LifecycleEventArgs` as type declaration here, as there are more implementations of `EventArgs` passed here (`Doctrine\ORM\Event\LoadClassMetadataEventArgs` by instance).

With that on mind, I guess we could use some type check and a deprecation path on these getters to prevent calling `$this->args->getObjectManager()` and `$this->args->getObject()` on objects with other type than `Doctrine\Persistence\Event\LifecycleEventArgs`.

I mean something like this:
```php
public function getObject(): object
{
    if (null === $this->args) {
        throw new \LogicException(sprintf('Event args must be set before calling "%s()".', __METHOD__));
    }

    if (!$this->args instanceof \Doctrine\Persistence\Event\LifecycleEventArgs) {
        @trigger_error(sprintf(
            'Calling "%s()" on an event that is not of type "%s" is deprecated since gedmo/doctrine-extensions 3.x'
            .' and will throw a "%s" error in version 4.0.',
            __METHOD__,
            LifecycleEventArgs::class,
            \TypeError::class
        ), E_USER_DEPRECATED);
    }

    // @todo Remove this check when `\Doctrine\Persistence\Event\LifecycleEventArgs` is used.
    if (\method_exists($this->args, 'getObject')) {
        return $this->args->getObject();
    }

    return $this->args->getEntity();
}

Please, confirm if my understanding about the current situation is right and let me know if I'm missing something.
Thank you.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: > From what I see, the case seems to be even worse than I thought, as the property used here (ORM::$args) is using a type (Doctrine\Common\EventArgs) that does not guarantee the getEntityManager() or the getEntityManager() methods

Correct, that's why these warnings are ignored for class in the php actions here.

> That said, I know we surely couldn't use LifecycleEventArgs as type declaration here, as there are more implementations of EventArgs passed here (Doctrine\ORM\Event\LoadClassMetadataEventArgs by instance).

I cannot confirm that is the case, but if it is, the current implementation would crash anyway because those EventArgs do not have a `getEntity` method.

From my understanding (and I could definitely be wrong there) all EventArgs actually passed from Doctrine to EventListeners are of type Persistence/LifecycleEventArgs. The only way I could see this breaking bc is if people implemented their own EventArgs implementations with getEntityManager and getEntity methods and these were somehow passed to our ORM Adapter.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: I'll just add some deprecation notices if the $args don't have a getObjectManager/getObject method and see if these are ever triggered to confirm. Give me a few days to test it.

Comment (phansys) [reply src/Mapping/Event/Adapter/ORM.php:98]: Thank you very much for the time and effort put on this!

In the test I used, I can confirm other types than `Doctrine\Persistence\Event\LifecycleEventArgs` are passed to this class, like `Doctrine\ORM\Event\LoadClassMetadataEventArgs`.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: Time to figure out why that doesn't cause a crash then. `getEntity` must just never be called on those.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:98]: Edit: actually, being on my pc, there's a very simple explanation for why that would still be okay:
https://github.com/doctrine/orm/blob/2.14.x/lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php

File: tests/Gedmo/Mapping/MappingEventAdapterTest.php
Comment (phansys) [inline tests/Gedmo/Mapping/MappingEventAdapterTest.php:55]: Maybe we could even replace these mocks to use a class that is no deprecated, like `Doctrine\ORM\Event\PrePersistEventArgs` or `Doctrine\ORM\Event\PreUpdateEventArgs`.

File: tests/Gedmo/Mapping/MappingEventAdapterTest.php
Comment (phansys) [reply tests/Gedmo/Mapping/MappingEventAdapterTest.php]: Regarding `getEventAdapter()`, I stumbled with that yesterday in my tests :disappointed_relieved: 

>GitHub doesn't have the correct emoji to express my feelings about that code. Sometimes I feel like code reflection goes just a little too far in some languages, hahaha.

:dizzy_face: :face_with_spiral_eyes: :rofl: 

>It would be awesome to be able to get rid of this code, but since it doesn't seem easy, maybe we should keep using the deprecated Doctrine\Persistence\Event\LifecycleEventArgs;, once https://github.com/doctrine-extensions/DoctrineExtensions/pull/2669 is merged we can use one of the new dedicated events.

Just for the records, `Doctrine\ORM\Event\LifecycleEventArgs` is deprecated, but `Doctrine\Persistence\Event\LifecycleEventArgs` is not.

**Comment by DubbleClick** [comment]
I didn't run into any deprecation triggers in the last two days with the code of this PR, although I'm only using Loggable, Timestampable and Blameable.

**Review by franmomu — APPROVED** [review]
Thanks @DubbleClick! sorry for the back and forth, please add a changelog entry

**Comment by DubbleClick** [comment]
Hi, sorry but I can't access a pc until next week. Will do then.

File: src/Mapping/Event/Adapter/ORM.php
Comment (franmomu) [inline src/Mapping/Event/Adapter/ORM.php:107]: Should this be `\Error::class`?

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php:107]: I think a TypeError is fine if we look at it as the $args class (type) not implementing the getObjectManager method. If we look at it like "the type $args::class must implement getObjectManager for version 4.0" then it should be \Error.

**Comment by DubbleClick** [comment]
@franmomu I've added the changelog note and changed the \TypeError to \Error

File: CHANGELOG.md
Comment (phansys) [inline CHANGELOG.md:24]:
```suggestion
- Calling `Gedmo\Mapping\Event\Adapter\ORM::getObjectManager()` and `getObject()` on EventArgs that do not implement `getObjectManager` and `getObject()` (such as old EventArgs implementing `getEntityManager()` and `getEntity()`) 
```

File: CHANGELOG.md
Comment (phansys) [inline CHANGELOG.md:25]:
```suggestion
- Calling `Gedmo\Uploadable\Event\UploadableBaseEventArgs::getEntityManager()` and `getEntity()`. Call `getObjectManager()` and `getObject()` instead. 
```

File: CHANGELOG.md
Comment (phansys) [inline CHANGELOG.md:22]: This will be added when a new release is created.
```suggestion
```

File: src/Mapping/Event/Adapter/ORM.php
Comment (phansys) [reply src/Mapping/Event/Adapter/ORM.php]: If there won't be an `instanceof` check here, I guess we can't even be sure if`\Error` will be thrown. If that is the case, I guess we should say something like "it will be not supported" instead.

**Comment by raziel057** [comment]
Hi, just for information I switched to the https://github.com/DubbleClick/DoctrineExtensions/commits/main branch on my project and works as expected. I still get deprecation related to the use of ``Doctrine\Persistence\Event\LifecycleEventArgs``. That should be fixed by merge of the following PR: https://github.com/doctrine-extensions/DoctrineExtensions/pull/2649

**Comment by DubbleClick** [comment]
@phansys "If there won't be an instanceof check here, I guess we can't even be sure if\Error will be thrown. If that is the case, I guess we should say something like "it will be not supported" instead."

No instanceof check is required, calling a method that doesn't exist throws an \Error.

`Fatal error: Uncaught Error: Call to undefined method OldEventArgs::getObjectManager()`

@raziel057 thanks for the confirmation, but be warned that the fork will likely be deleted the day this is merged :p

File: src/Mapping/Event/Adapter/ORM.php
Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php]: No instanceof check is required, calling a method that doesn't exist throws an \Error.

Fatal error: Uncaught Error: Call to undefined method OldEventArgs::getObjectManager()

**Comment by raziel057** [comment]
@DubbleClick No problem thanks we will switch back as soon as this is merged.

**Comment by phansys** [comment]
>@phansys "If there won't be an instanceof check here, I guess we can't even be sure if\Error will be thrown. If that is the case, I guess we should say something like "it will be not supported" instead."
>No instanceof check is required, calling a method that doesn't exist throws an \Error.

What is strange (and wrong) at first glance to me, is that we will call a method that is not guaranteed by the object we are trying to call it from. In other words, we are allowing a method call without confirming whether that is possible.
I'd like to understand what you think the method's body will be like in the next major release, because I guess we should avoid calling a non-existent method, preventing the call with our own mechanism in the worst case if we can't trust the contract provided by the class or interface.

**Comment by DubbleClick** [comment]
> What is strange (and wrong) at first glance to me, is that we will call a method that is not guaranteed by the object we are trying to call it from. In other words, we are allowing a method call without confirming whether that is possible. I'd like to understand what you think the method's body will be like in the next major release, because I guess we should avoid calling a non-existent method, preventing the call with our own mechanism in the worst case if we can't trust the contract provided by the class or interface.

```php
if (\method_exists($this->args, 'getObject')) {
    return $this->args->getObject();
}

@trigger_error(sprintf(
    'Calling "%s()" on an event of type "%s" is deprecated since gedmo/doctrine-extensions 3.x'
    .' and will throw a "%s" error in version 4.0.',
    __METHOD__,
    get_class($this->args),
    \Error::class
), E_USER_DEPRECATED);
```

becomes 

```php
return $this->args->getObject();
```

If the method doesn't exist, PHP throws an \Error.

**Comment by phansys** [comment]
>If the method doesn't exist, PHP throws an \Error.
Just like Doctrine will result in the same error when you attempt to call getEntity() on their EventArgs next release.

Thanks for the clarification.
I don't feel comfortable admitting an internal PHP error (`\Error`) when we already know this problem is part of a design flaw on our side.

**Comment by DubbleClick** [comment]
> > If the method doesn't exist, PHP throws an \Error.
> > Just like Doctrine will result in the same error when you attempt to call getEntity() on their EventArgs next release.
> 
> Thanks for the clarification. I don't feel comfortable admitting an internal PHP error (`\Error`) when we already know this problem is part of a design flaw on our side.

I don't think it's a design flaw - doctrine changed its functionality to deprecate `getEntity()` and replace it by `getObject()`, so we react to it and call `getObject()`. For the current version we keep `getEntity()` as a fallback for people using their own EventArgs implementations, but deprecate it. In the next major version, following Doctrine, we can expect people to have refactored their own EventArgs.

File: src/Mapping/Event/Adapter/ORM.php
Comment (franmomu) [reply src/Mapping/Event/Adapter/ORM.php]: Can we maybe add instanceof checks with `LifecycleEventArgs`, `LoadClassMetadataEventArgs`, `ManagerEventArgs` and `OnClearEventArgs`? Maybe they should all extend from `ManagerEventArgs`  in `doctrine/persistence` 🤔 ? For the `getObjectManager()` call I mean, or any other solution?

It would be nice to finally move this forward.

Comment (DubbleClick) [reply src/Mapping/Event/Adapter/ORM.php]: I don't see any need to do that. The \method_exists call will make sure that it's fully backwards compatible for now.

If anything, that would be a discussion for later when we remove the  \method_exists call and directly call `getObjectManager()`. But even then, I don't see a point in checking the type of the args - all we care about is that they abide by Doctrines interfaces and implement `getObjectManager()` instead of `getEntityManager()` (that we currently also call without checking).

**Comment by franmomu** [comment]
> > If the method doesn't exist, PHP throws an \Error.
> > Just like Doctrine will result in the same error when you attempt to call getEntity() on their EventArgs next release.
> 
> Thanks for the clarification. I don't feel comfortable admitting an internal PHP error (`\Error`) when we already know this problem is part of a design flaw on our side.

@phansys so what do you think we should do here, it would be nice to finally finish and merge this PR.

**Comment by phansys** [comment]
>@phansys so what do you think we should do here, it would be nice to finally finish and merge this PR.

First of all, sorry for the delay on the feedback.

Currently, the PR has some deprecation messages that indicate what will be disallowed in the next major release, but IMO, there is no clear path about how the restriction will be implemented.

```php
    public function getObject(): object
    {
        if (null === $this->args) {
            throw new \LogicException(sprintf('Event args must be set before calling "%s()".', __METHOD__));
        }

        if (\method_exists($this->args, 'getObject')) {
            return $this->args->getObject();
        }

        @trigger_error(sprintf(
            'Calling "%s()" on an event of type "%s" is deprecated since gedmo/doctrine-extensions 3.x'
            .' and will throw a "%s" error in version 4.0.',
            __METHOD__,
            get_class($this->args),
            \Error::class
        ), E_USER_DEPRECATED);

        return $this->args->getEntity();
    }
```

By instance, I don't know if the call to `method_exists()` will be removed because we should be trusting in the type of `$this->args` based on a previous type-safe check, or if we are proposing to let the PHP engine to reach a condition where we call a non-existent method `getObject()` which can be not defining depending on the implementation of `$this->args`.

I think a comment with the code that must replace these deprecations should be enough.
I'm thinking of a kind of instruction like this:
https://github.com/doctrine-extensions/DoctrineExtensions/blob/aaa8c15645df41c6247b039cade29948162e27f6/src/Mapping/Driver/Chain.php#L90-L92

**Comment by DubbleClick** [comment]
> By instance, I don't know if the call to method_exists() will be removed because we should be trusting in the type of $this->args based on a previous type-safe check, or if we are proposing to let the PHP engine to reach a condition where we call a non-existent method getObject() which can be not defining depending on the implementation of $this->args.

Isn't that necessarily the same behaviour as it is now? There's even less of a guarantee that the EventArgs implement `getEntityManager` than there is that they implement `getObjectManager`.

I'm happy to write a todo block for the next major release, though.

**Comment by phansys** [comment]
>Isn't that necessarily the same behaviour as it is now? There's even less of a guarantee that the EventArgs implement getEntityManager than there is that they implement getObjectManager.

If this is the case, I'd choose a path that leaves under our own control the responsibility to throw an error explaining in a clear way why the configured `$this->args` is not valid for the context where it is used.

**Comment by DubbleClick** [comment]
I added instructions and cleared up the error message a little better.

Honestly, I am not even sure how far out of their way people would have to go to have DoctrineExtensions catch their own EventArgs implementations. All the Doctrine events we catch and our own implement `getObjectManager`.

**Review by phansys — APPROVED** [review]

**Comment by franmomu** [comment]
Thanks @DubbleClick ❤️

**Comment by brzuchal** [comment]
When to expect a release of this?

**Comment by DubbleClick** [comment]
in all likelihood before orm 3.0, so I wouldn't worry about it :)

**Comment by franmomu** [comment]
I hope to release this soon, but I would like to include in the release also these PR:
https://github.com/doctrine-extensions/DoctrineExtensions/pull/2715
https://github.com/doctrine-extensions/DoctrineExtensions/pull/2716
https://github.com/doctrine-extensions/DoctrineExtensions/pull/2720
https://github.com/doctrine-extensions/DoctrineExtensions/pull/2556

**Pull request merged**
